### ISSUE TITLE
Should `loss` be divided by `gradient_accumulation_steps`?

### ISSUE THREAD

**Issue by wjfwzzc** [karpathy/nanoGPT#168]
@karpathy  I notice that `loss` is NOT divided by `gradient_accumulation_steps` before backward, but should the loss be averaged across the whole batch size?
https://github.com/karpathy/nanoGPT/blob/ae3a8d5fdd3ddb8b13fab182723476523961e3ab/train.py#L281-L293

**Comment by karpathy** [comment]
You're right it should be, I'll issue a fix. One thing to note is that this is less of a bug than it appears to be because Adam is scale invariant, so multiplying the scale by any constant is a no-op - you'll get the same optimization except your final loss will be scaled. Other parts of the loop would be impacted though, e.g. especially gradient clipping. TLDR I think the code as is produces the desired optimization, except our gradient clip value is silently lower by a factor of `gradient_accumulation_steps`.

**Comment by wjfwzzc** [comment]
I see your point. 

> because Adam is scale invariant, so multiplying the scale by any constant is a no-op

Though I think it's not entirely scale invariant due to the $\epsilon$ in the denominator. In my experience the affect of the $\epsilon$ may not be negligible, especially in bf16/fp16 mode or on some noisy datasets :(
